# for each TCGA sample, result is generated by the following command
sample_1_fastq= # fastq file for sample read1
sample_2_fastq= # fastq file for sample read2
StarIndex= # full path folder for STAR index
sample_Output= # output folder
GTFfile= # path for GTF annotation file


# STAR alignment
STAR --runThreadN 16 --genomeDir ${StarIndex} --readFilesIn ${sample_1_fastq} ${sample_2_fastq} --outFileNamePrefix ${sample_Output} --outSAMtype BAM SortedByCoordinate -outSAMstrandField intronMotif --chimSegmentMin 15
samtools view -Shb ${sample_Output}/Chimeric.out.sam -o ${sample_Output}/Chimeric.out.bam

# remove chimeric reads if they can be aligned concordantly by speedseq
samtools fastq -1 ${sample_Output}/ChimRead_1.fastq -2 ${sample_Output}/ChimRead_2.fastq ${sample_Output}/Chimeric.out.bam
mkdir -p ${sample_Output}/speedseqChim
speedseq align -R "@RG\tID:id\tSM:samplename\tLB:lib" -o ${sample_Output}/speedseqChim/Aligned ${FASTAFile} ${sample_Output}/ChimRead_1.fastq ${sample_Output}/ChimRead_2.fastq
python3 FilterChimOut.py ${sample_Output}/speedseqChim/Aligned.bam ${sample_Output}/Chimeric.out.bam ${sample_Output}/Chimeric.out.filtered.bam

# squid
squid -b ${sample_Output}/Aligned.sortedByCoord.out.bam -c ${sample_Output}/Chimeric.out.filtered.bam -r 1 -o ${sample_Output}/squid

# post-processing
awk '{if(!($1==14 && $4==14 && substr($2,0,2)=="10" && substr($5,0,2)=="10") && !($1==22 && $4==22 && substr($2,0,2)=="22" && substr($5,0,2)=="22") && !($1==2 && $4==2 && substr($2,0,1)=="8" && substr($5,0,1)=="8")) print $0}' ${sample_Output}/squid_sv.txt > ${sample_Output}/sqfilter_sv.txt;

# annotate TSV with gtf file, fusion-gene and non-fusion-gene label is generated by this process
python AnnotSV.py 1 ${sample_Output}/sqfilter_sv.txt ${GTFfile} ${sample_Output}/sqfilter_sv_Annot.txt
